src/plugins/shell/shell_impl.cpp

```diff
@@ -0,0 +1,145 @@
+#include "shell_impl.h"
+#include "system.h"
+#include "global_include.h"
+#include "px4_custom_mode.h"
+#include <cmath>
+#include <functional>
+#include <array>
+
+namespace mavsdk {
+
+void ShellImpl::init()
+{
+    using namespace std::placeholders; // for `_1`
+    _parent->register_mavlink_message_handler(
+        MAVLINK_MSG_ID_SERIAL_CONTROL,
+        std::bind(&ShellImpl::process_shell_message, this, _1),
+        this);
+}
+
+void ShellImpl::deinit() {}
+
+void ShellImpl::enable() {}
+
+void ShellImpl::disable() {}
+
+ShellImpl::ShellImpl(System& system) : PluginImplBase(system)
+{
+    _parent->register_plugin(this);
+}
+
+ShellImpl::~ShellImpl()
+{
+    _parent->unregister_plugin(this);
+}
+
+void ShellImpl::shell_message_response_async(Shell::result_callback_t callback)
+{
+    _result_subscription = callback;
+}
+
+void ShellImpl::send_shell_message(const Shell::ShellMessage& shell_message)
+{
+    if (!_parent->is_connected()) {
+        if (_result_subscription) {
+            auto callback = _result_subscription;
+            Shell::Result arg{Shell::Result::ResultCode::NO_SYSTEM, {}};
+            _parent->call_user_callback([callback, arg]() { callback(arg); });
+            return;
+        }
+    }
+    if (shell_message.data.length() > 70 && _result_subscription) {
+        auto callback = _result_subscription;
+        Shell::Result arg{Shell::Result::ResultCode::DATA_TOO_LONG, {}};
+        _parent->call_user_callback([callback, arg]() { callback(arg); });
+        return;
+    }
+
+    _shell_message_mutex.lock();
```

Comment:
I suggest to always use `std::lock_guard` for anything that we add now, and also remove the old `lock()`, `unlock()`s at some point.

```
{
    std::lock_guard<std::mutex> lock(_shell_message_mutex);
    _shell_message = shell_message;
}
```